src/main/java/org/wildfly/security/auth/realm/token/validator/JwkManager.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2018 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.wildfly.security.auth.realm.token.validator;
+
+import javax.json.Json;
+import javax.json.JsonArray;
+import javax.json.JsonObject;
+import javax.net.ssl.HostnameVerifier;
+import javax.net.ssl.HttpsURLConnection;
+import javax.net.ssl.SSLContext;
+import java.io.IOException;
+import java.io.InputStream;
+import java.math.BigInteger;
+import java.net.URL;
+import java.net.URLConnection;
+import java.security.KeyFactory;
+import java.security.NoSuchAlgorithmException;
+import java.security.PublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.RSAPublicKeySpec;
+import java.util.Base64;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import static org.wildfly.security._private.ElytronMessages.log;
+
+/**
+ * Object for caching json web keys for signature validation
+ * @author <a href="mailto:mmazanek@redhat.com">Martin Mazanek</a>
+ */
+class JwkManager {
+
+    private final Map<URL, Map<String, RSAPublicKey>> keys;
+    private final Map<URL, Long> timeouts;
+    private final SSLContext sslContext;
+    private final HostnameVerifier hostnameVerifier;
+
+    private final long updateTimeout;
+
+    private static final int CONNECTION_TIMEOUT = 2000;//2s
+
+    JwkManager(SSLContext sslContext, HostnameVerifier hostnameVerifier, long updateTimeout) {
+        keys = new LinkedHashMap<>();
+        timeouts = new LinkedHashMap<>();
+        this.sslContext = sslContext;
+        this.hostnameVerifier = hostnameVerifier;
+        this.updateTimeout = updateTimeout;
+    }
+
+    public PublicKey getPublicKey(String kid, URL url) {
```

Comment:
I assume there is one instance of this class which may be used by multiple threads?  If so likely going to need to consider the synchronisation requirements as it currently has reads and updates.

Also probably should have some javadoc detailing the concurrency requirements of this class.